### PR TITLE
Frontend ducks

### PR DESCRIPTION
**Pull request by Gailbear**

### CODE REVIEW

File: frontend/src/app/_components/duck/duck.component.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+})
+export class DuckComponent implements OnInit {
+  @Input() duckId!: number;
+  duck$!: Observable<Duck|undefined>;
+
+  constructor(private duckService: DuckService,
+    private router: Router) { }
+
+  ngOnInit(): void {
+    console.log('duck is init!')
```
Comment (Gailbear) [inline frontend/src/app/_components/duck/duck.component.ts:21]: note to self, clean this up

File: frontend/src/app/_services/duck.service.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  get_ducks() : Observable<Duck[]> {
+    return this.http.get<Duck[]>(`${environment.backend}/ducks`);
+  }
+
+  get_duck(id: number) : Observable<Duck | undefined> {
+    return this.http.get<Duck|undefined>(`${environment.backend}/ducks/${id}`);
+  }
+
+  award_duck(id: number) {
+    return this.http.post<any>(`${environment.backend}/ducks/${id}`, {}, {withCredentials: true}).pipe(
```
Comment (clahey) [inline frontend/src/app/_services/duck.service.ts:26]: Why does this return an any?

File: frontend/src/app/_services/duck.service.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      catchError(e => of(false))
```
Comment (clahey) [inline frontend/src/app/_services/duck.service.ts:27]: Why false instead of undefined?

File: frontend/src/app/_services/duck.service.ts (hunk 2 above)
Comment (clahey) [inline frontend/src/app/_services/duck.service.ts:26]: This Observable should probably include a shareReplay() in the pipe and should be subscribed to here in case the caller doesn't subscribe to it. See account.service.ts for an example.

File: arisia-remote/app/arisia/fun/DuckService.scala [hunk 4]
```diff
@@ -99,7 +99,7 @@ class DuckServiceImpl(
           )
         } else {
           // Somehow got a request from the wrong place, which seems hinky:
-          val error = s"Got a request for Duck $duckId from incorrect location $from!"
+          val error = s"Got a request for Duck $duckId from incorrect location $from! It's supposed to be ${duck.requestingUrl}"
```
Comment (clahey) [inline arisia-remote/app/arisia/fun/DuckService.scala:102]: Should we not include the correct URL? If someone does guess the duck, but isn't at the right place, won't their network tab tell them where to go?

Comment (clahey) [inline arisia-remote/app/arisia/fun/DuckService.scala:102]: We could also make the duck ids be 8 digit random strings so they can't be guessed as easily.

File: frontend/src/app/user/duck-list/duck-list.component.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  @Input() duckIds: number[] = [];
+  @Input() isSelf: boolean = false;
+
+
+  constructor(private duckService: DuckService) { }
+
+  ngOnInit(): void {
+    this.ducks$ = this.duckService.get_ducks().pipe(
+      map( ducks => ducks.map(duck => {
+        duck.hidden = !this.duckIds.includes(duck.id);
```
Comment (clahey) [inline frontend/src/app/user/duck-list/duck-list.component.ts:23]: We shouldn't be updating the ducks in the service based on the ids in the component page. Why not have the duck service know about which ducks the user has and just mark the ones they don't have as hidden when downloading the list from the server? And then when awarding ducks, update that value?

It can just inject account.service to get that info.

File: frontend/src/app/user/user.component.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import { Component, OnInit } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
 import { Observable } from 'rxjs';
-import { pluck, switchMap } from 'rxjs/operators';
+import { pluck, switchMap, tap } from 'rxjs/operators';
```
Comment (clahey) [inline frontend/src/app/user/user.component.ts:4]: Why this change?

File: frontend/src/app/_services/account.service.ts [hunk 7]
```diff
@@ -59,13 +59,22 @@ export class AccountService {
...
+      this.user.ducks = [1];
       return of(this.user);
     } else {
-      // This is for testing.
-      return of({id: 'joe', name: 'Joe', badgeNumber: badgeNumber, zoomHost: false});
-      // below is the real one
-      // this.http.get<User>(`${environment.backend}/user/${id}`, {withCredentials: true})
+      return of({id: 'joe', name: 'Joe', badgeNumber: badgeNumber, zoomHost: false, ducks: [1,3], self: false});
     }
+    // below is the real one
```
[the remark this answers is not in the document]
Comment (Gailbear) [reply frontend/src/app/_services/account.service.ts:72]: ducks and ribbons, and we don't re-fetch me..

File: arisia-remote/app/arisia/fun/DuckService.scala (hunk 4 above)
Comment (Gailbear) [reply arisia-remote/app/arisia/fun/DuckService.scala:102]: oooh good call, i had put that in there for debugging but i don't want that

Comment (Gailbear) [reply arisia-remote/app/arisia/fun/DuckService.scala:102]: :( nope backend restriction at the moment

File: frontend/src/app/user/duck-list/duck-list.component.ts (hunk 5 above)
Comment (Gailbear) [reply frontend/src/app/user/duck-list/duck-list.component.ts:23]: hmm true.

File: frontend/src/app/user/user.component.ts (hunk 6 above)
Comment (Gailbear) [reply frontend/src/app/user/user.component.ts:4]: unintentional from logging!

File: frontend/src/app/user/duck-list/duck-list.component.ts (hunk 5 above)
Comment (Gailbear) [reply frontend/src/app/user/duck-list/duck-list.component.ts:23]: but what about other people's ducks? imean i guess i can still just do it through the ducks endpoint...
